vault/resource_ad_secret_library.go

```diff
@@ -0,0 +1,232 @@
+	log.Printf("[DEBUG] Wrote %q", setPath)
+	return readLibraryResource(d, meta)
+}
+
+func readLibraryResource(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	setPath := d.Id()
+	log.Printf("[DEBUG] Reading %q", setPath)
+
+	setName, err := adSecretBackendSetNameFromPath(setPath)
+	if err != nil {
+		log.Printf("[WARN] Removing AD library %q because its ID is invalid", setPath)
+		d.SetId("")
+		return fmt.Errorf("invalid library ID %q: %s", setPath, err)
+	}
+	d.Set("name", setName)
+
+	backend, err := adSecretBackendFromLibraryPath(setPath)
+	if err != nil {
+		log.Printf("[WARN] Removing AD library %q because its ID is invalid", setPath)
+		d.SetId("")
+		return fmt.Errorf("invalid library ID %q: %s", setPath, err)
+	}
+	d.Set("backend", backend)
+
+	resp, err := client.Logical().Read(setPath)
+	if err != nil {
+		return fmt.Errorf("error reading %q: %s", setPath, err)
+	}
+	log.Printf("[DEBUG] Read %q", setPath)
+
+	if resp == nil {
+		log.Printf("[WARN] %q not found, removing from state", setPath)
+		d.SetId("")
+		return nil
+	}
+
+	if val, ok := resp.Data["service_account_names"]; ok {
+		if err := d.Set("service_account_names", val); err != nil {
+			return fmt.Errorf("error setting state key 'service_account_names': %s", err)
+		}
+	}
+
+	if val, ok := resp.Data["max_ttl"]; ok {
+		if err := d.Set("max_ttl", val); err != nil {
+			return fmt.Errorf("error setting state key 'max_ttl': %s", err)
+		}
+	}
+
+	if val, ok := resp.Data["ttl"]; ok {
+		if err := d.Set("ttl", val); err != nil {
+			return fmt.Errorf("error setting state key 'ttl': %s", err)
+		}
+	}
+
+	if val, ok := resp.Data["disable_check_in_enforcement"]; ok {
+		if err := d.Set("disable_check_in_enforcement", val); err != nil {
+			return fmt.Errorf("error setting state key 'disable_check_in_enforcement': %s", err)
+		}
+	}
+
+	return nil
+}
+
+func updateLibraryResource(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	setPath := d.Id()
+	log.Printf("[DEBUG] Updating %q", setPath)
+
+	data := map[string]interface{}{}
+	if raw, ok := d.GetOk("service_account_names"); ok {
+		data["service_account_names"] = raw
+	}
+
+	if raw, ok := d.GetOk("max_ttl"); ok {
+		data["ttl"] = raw
+	}
+
+	if raw, ok := d.GetOk("ttl"); ok {
+		data["ttl"] = raw
+	}
+
+	if raw, ok := d.GetOk("disable_check_in_enforcement"); ok {
+		data["disable_check_in_enforcement"] = raw
+	}
+
+	if _, err := client.Logical().Write(setPath, data); err != nil {
+		return fmt.Errorf("error updating library %q: %s", setPath, err)
+	}
+	log.Printf("[DEBUG] Updated %q", setPath)
+	return readLibraryResource(d, meta)
+}
+
+func deleteLibraryResource(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	setPath := d.Id()
+	log.Printf("[DEBUG] Deleting %q", setPath)
+
+	if _, err := client.Logical().Delete(setPath); err != nil && !util.Is404(err) {
+		return fmt.Errorf("error deleting %q: %s", setPath, err)
```

Comment:
404 check here?

Reply:
Unless deleting a thing that doesn't exist doesn't return an error 🤔

Reply:
It doesn't, just a 204:

```
curl -I \
    --header "X-Vault-Token: root" \
    --request DELETE \
    http://127.0.0.1:8200/v1/ad/library/clint
HTTP/1.1 204 No Content
Cache-Control: no-store
Content-Type: application/json
Date: Tue, 17 Nov 2020 00:13:09 GMT
```